state/runtime/system/system_unstaking.go

```diff
@@ -26,30 +27,50 @@ func (uh *unstakingHandler) run(state *systemState) ([]byte, error) {
 	// Grab the address calling the staking method
 	staker := state.contract.Caller
 
-	// Grab the staked balance for the account
+	// Grab the last recorded staked balance for the account
 	stakedBalance := state.host.GetStakedBalance(staker)
 
 	// Grab the balance on the staking address
 	stakingAccountBalance := state.host.GetBalance(stakingAddress)
 
+	// Grab the transaction context
+	ctx := state.host.GetTxContext()
+
+	// Construct the pending event
+	pendingEvent := staking.PendingEvent{
+		BlockNumber: ctx.Number,
+		Address:     staker,
+		Value:       big.NewInt(0),
+		EventType:   staking.UnstakingEvent,
+	}
+
+	// Find the staked balance after any possible events occur
+	afterEventsStake := state.host.ComputeStakeAfterEvents(stakedBalance, pendingEvent)
+	zeroValue := big.NewInt(0)
+
 	// Sanity check
 	// Can't unstake if the balance isn't previously present on the staking account
 	// Can't unstake if the value is different from 0
-	if stakingAccountBalance.Cmp(stakedBalance) < 0 || state.contract.Value.Cmp(big.NewInt(0)) != 0 {
+	// Can't unstake if the account doesn't have stake
+	// Can't unstake if the account has pending events which cause it to not have stake
+	if stakingAccountBalance.Cmp(afterEventsStake) < 0 ||
+		state.contract.Value.Cmp(zeroValue) != 0 ||
+		stakedBalance.Cmp(zeroValue) == 0 ||
+		afterEventsStake.Cmp(zeroValue) == 0 {
```

Comment:
I think that you need to omit this check (`stakedBalance.Cmp(zeroValue) == 0`):
```suggestion
	if stakingAccountBalance.Cmp(afterEventsStake) < 0 ||
		state.contract.Value.Cmp(zeroValue) != 0 ||
		afterEventsStake.Cmp(zeroValue) == 0 {
```

If you do not remove it, the following scenario becomes impossible:
```
Prior to the block user has a balance of 5 ETH and staked balance of 0 ETH.
In the same block he does the following:

1. Stake 2 ETH
2. Unstake
```

If you agree, please add this test scenario to `pos_test.go`. It isn't an exploit, so maybe something like `TestPoS_StakeUnstakeWithinSameBlock`

Reply:
You're right 👍 
I changed the logic below, but forgot to remove the stale `stakedBalance` from the check. We already have `afterEventsStake` which is the most up to date staked balance.

Changing now, and adding a test 👍

Reply:
Good job on the test, it's on point 🎯